api/api/routers/mcp/_mcp_observability_session_state.py

```diff
@@ -0,0 +1,193 @@
+        """Load session state from Redis"""
+        if not shared_redis_client:
+            logger.warning("Redis client not available, cannot load session state")
+            return None
+
+        try:
+            redis_key = cls._get_redis_key(tenant, session_id)
+            session_data_str = await shared_redis_client.get(redis_key)
+
+            if not session_data_str:
+                logger.debug(
+                    "Session not found in Redis",
+                    extra={"session_id": session_id},
+                )
+                return None
+
+            session_data = json.loads(session_data_str.decode())
+
+            # Reconstruct session state
+            session = cls.__new__(cls)
+            session.session_id = session_data["session_id"]
+            session.created_at = datetime.fromisoformat(session_data["created_at"])
+            session.last_activity = datetime.fromisoformat(session_data["last_activity"])
+            session.tool_calls = session_data["tool_calls"]
+
+            logger.debug(
+                "Session loaded from Redis",
+                extra={
+                    "session_id": session_id,
+                    "tool_calls_count": len(session.tool_calls),
+                    "created_at": session.created_at,
+                },
+            )
+
+            return session
+
+        except Exception as e:
+            logger.error(
+                "Failed to load session from Redis",
+                extra={
+                    "session_id": session_id,
+                    "error": str(e),
+                },
+            )
+            return None
+
+    @classmethod
+    async def get_or_create(cls, session_id: str | None, tenant: TenantData) -> tuple["SessionState", bool]:
+        """
+        Get existing session or create a new one.
+
+        Returns:
+            tuple: (SessionState, is_new_session)
+        """
+        # If no session ID provided or session doesn't exist, create new one
+        if session_id:
+            if found_session := await cls.load(session_id, tenant):
+                # Update last activity and extend expiration
+                found_session.last_activity = datetime.now(timezone.utc)
+                await found_session.save(tenant)
+                return found_session, False
+
+        # Create new session
+        new_session_id = str(uuid.uuid4())
+        new_session = cls(new_session_id)
+        await new_session.save(tenant)
+
+        logger.info(
+            "New MCP session created",
+            extra={
+                "mcp_event": "session_created",
+                "session_id": new_session_id,
+                "session_created_at": new_session.created_at,
+                "replaced_session_id": session_id,
+            },
+        )
+
+        return new_session, True
+
+    async def register_tool_call(self, tool_call_data: ToolCallData, tenant: TenantData) -> None:
```

Comment:
the tenant should be a SessionState attribute ? The current functions suggest that it is possible to update for a single session ?